src/meadowrun/aws_integration/grid_tasks_sqs.py

```diff
@@ -356,41 +410,31 @@ async def get_results_unordered(
     # implementation only relies on S3 instead, which allows faster and more concurrent
     # downloads.
 
-    # TODO monitor workers and react appropriately if a worker crashes.
-    if workers_done is None:
-        workers_done = asyncio.Event()
-
-    num_tasks_completed = 0
+    # num_tasks_completed = 0
     results_prefix = _s3_results_prefix(job_id)
     download_keys_received: Set[str] = set()
     wait = True
-    workers_done_wait_count = 0
+    workers_exited_wait_count = 0
     session = aiobotocore.session.get_session()
     async with session.create_client("s3", region_name=region_name) as s3c:
-        while num_tasks_completed < num_tasks and workers_done_wait_count < 3:
+        while not stop_receiving.is_set() and workers_exited_wait_count < 3:
 
-            if workers_done.is_set():
-                print(
-                    "All workers exited, waiting for task results. "
-                    f"Requested: {num_tasks}, received: {num_tasks_completed}"
-                )
-                workers_done_wait_count += 1
-            else:
-                print(
-                    f"Waiting for task results. Requested: {num_tasks}, "
-                    f"received: {num_tasks_completed}"
-                )
+            if all_workers_exited.is_set():
+                workers_exited_wait_count += 1
 
             if wait:
-                try:
-                    await asyncio.wait_for(
-                        workers_done.wait(),
-                        timeout=1.0  # poll more frequently if workers are done.
-                        if workers_done.is_set()
-                        else receive_message_wait_seconds,
-                    )
-                except asyncio.TimeoutError:
-                    pass
+                # poll more frequently if workers are done.
+                timeout = (
+                    1.0 if all_workers_exited.is_set() else receive_message_wait_seconds
+                )
+                done, pending = await asyncio.wait(
+                    [stop_receiving.wait(), all_workers_exited.wait()],
+                    timeout=timeout,
```

Comment:
@kurtschelfthout I think this should have `return_when=asyncio.FIRST_COMPLETED` I can include in my next change if you agree?